sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -5237,6 +5237,59 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
       )
     )
   }
+
+  test("SPARK-41232 array_append -> Unit Test cases for the function ") {
+    val df1 = Seq((Array[Int](3, 2, 5, 1, 2), 3)).toDF("a", "b")
```

Comment:
I want to know if `array` is `Array[Double](3.0, 2.0)` and `element` is `Int 1`, what is the result, will it cast and return `Array[Double](3.0, 2.0,1.0)` or throw an exception?

Reply:
It will throw an exception, since in the input types we are trying to match the exact element. Let me add a unit test for the same as well.

Reply:
BTW I added another test in the DataFrameSuite and it is promoting the element to double and then appending it as a double element in the array.

Reply:
But for an existing function, such as `ArrayUnion`:

```
val df = Seq((Array(1.0, 2.0, 3.0), Array(4, 2))).toDF("a", "b")
val rows = df.select(array_union($"a", $"b")).collect()
```
the `rows` will be `[WrappedArray(1.0, 2.0, 3.0, 4.0)]` 

So, I think `array_append` should support similar cat behavior.

Reply:
good question, I try with other build-functions, and think we should follow `array_remove` in this case:

```
scala> spark.sql("""SELECT array_remove(a,b) FROM VALUES (ARRAY(1, 2, 3), 3) AS tab(a, b)""").show
+------------------+
|array_remove(a, b)|
+------------------+
|            [1, 2]|
+------------------+


scala> spark.sql("""SELECT array_remove(a,b) FROM VALUES (ARRAY(1.0, 2.0, 3.0), 3) AS tab(a, b)""").show
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.ARRAY_FUNCTION_DIFF_TYPES] Cannot resolve "array_remove(a, b)" due to data type mismatch: Input to `array_remove` should have been "ARRAY" followed by a value with same element type, but it's ["ARRAY<DECIMAL(2,1)>", "INT"].; line 1 pos 7;
'Project [unresolvedalias(array_remove(a#122, b#123), None)]
+- SubqueryAlias tab
   +- LocalRelation [a#122, b#123]

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.dataTypeMismatch(package.scala:73)
...

scala> spark.sql("""SELECT array_remove(a,b) FROM VALUES (ARRAY(1, 2, 3), 3.0) AS tab(a, b)""").show
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.ARRAY_FUNCTION_DIFF_TYPES] Cannot resolve "array_remove(a, b)" due to data type mismatch: Input to `array_remove` should have been "ARRAY" followed by a value with same element type, but it's ["ARRAY<INT>", "DECIMAL(2,1)"].; line 1 pos 7;
'Project [unresolvedalias(array_remove(a#124, b#125), None)]
+- SubqueryAlias tab
   +- LocalRelation [a#124, b#125]

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.dataTypeMismatch(package.scala:73)
...

scala> spark.sql("""SELECT array_position(a,b),a,b FROM VALUES (ARRAY(1, NULL, 3), 1.0) AS tab(a, b)""")
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.ARRAY_FUNCTION_DIFF_TYPES] Cannot resolve "array_position(a, b)" due to data type mismatch: Input to `array_position` should have been "ARRAY" followed by a value with same element type, but it's ["ARRAY<INT>", "DECIMAL(2,1)"].; line 1 pos 7;
'Project [unresolvedalias(array_position(a#564, b#565), None), a#564, b#565]
+- SubqueryAlias tab
   +- LocalRelation [a#564, b#565]

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.dataTypeMismatch(package.scala:73)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis0$5(CheckAnalysis.scala:249)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis0$5$adapted(CheckAnalysis.scala:236)
...

scala> spark.sql("""SELECT array_position(a,b),a,b FROM VALUES (ARRAY(1.0, NULL, 3.0), 1) AS tab(a, b)""")
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.ARRAY_FUNCTION_DIFF_TYPES] Cannot resolve "array_position(a, b)" due to data type mismatch: Input to `array_position` should have been "ARRAY" followed by a value with same element type, but it's ["ARRAY<DECIMAL(2,1)>", "INT"].; line 1 pos 7;
'Project [unresolvedalias(array_position(a#566, b#567), None), a#566, b#567]
+- SubqueryAlias tab
   +- LocalRelation [a#566, b#567]

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.dataTypeMismatch(package.scala:73)
...

scala> spark.sql("""SELECT array_contains(a,b),a,b FROM VALUES (ARRAY(1.0, NULL, 3.0), 1) AS tab(a, b)""")
res50: org.apache.spark.sql.DataFrame = [array_contains(a, b): boolean, a: array<decimal(2,1)> ... 1 more field]

scala> spark.sql("""SELECT array_contains(a,b),a,b FROM VALUES (ARRAY(1, NULL, 3), 1.0) AS tab(a, b)""")
res51: org.apache.spark.sql.DataFrame = [array_contains(a, b): boolean, a: array<int> ... 1 more field]
```

So currently behavior (exactly matching) looks reasonable to me.